### DISCUSSION TITLE
[Technical Questions]  Adding HL7 FHIR json to questionnaire

### DISCUSSION THREAD

**Discussion by carolinentran** [CS342/.github#47] [Technical Questions] [answered]
### In what area do you have a technical challenge?

Swift

### Description

I'm working on the CS342-Behavior project and my task is to modify an existing questionnaire. I have successfully modified the existing questionnaire content and downloaded the JSON for the questionnaire using HL7 FHIR. I'm reading through existing code right now and am having trouble figuring out where the JSON is populated into the questionnaire. 

I see that the current questionnaire is using the JSON in Behavior/Resources/SocialSupportQuestionnaire. I'm going through the ScheduleView code (which is where the questionnaire seems to be made viewable) but I'm not sure where the JSON is used. 

### Reproduction

I'm having trouble reading through the code and may have misunderstood some lines. Here is what I'm currently reading through: https://github.com/CS342/2024-Behavior/blob/365126e5eeb929646919dfb696aaa8369f2ea56d/Behavior/Schedule/ScheduleView.swift

### Expected behavior

Right now, the questionnaire is loading in the questions and contents of the Social Support Questionnaire. 
<img width="271" alt="Screenshot 2024-01-14 at 3 52 47 PM" src="https://github.com/CS342/.github/assets/89322465/77b39be5-0667-4109-845e-ced2cfac6815">

### Code of Conduct

- [X] I agree to follow this course's Code of Conduct and Contributing Guidelines

**Comment by vishnuravi** [comment, marked as answer]
Hi @carolinentran,

The questionnaire JSON is being imported in [BehaviorScheduler.swift](https://github.com/CS342/2024-Behavior/blob/main/Behavior/Schedule/BehaviorScheduler.swift) located in [Behavior/Schedule](https://github.com/CS342/2024-Behavior/tree/main/Behavior/Schedule).

If you look at lines 32-42, you'll see where the [Task](https://swiftpackageindex.com/stanfordspezi/spezischeduler/documentation/spezischeduler/task) that represents the Social Support Questionnaire is being created:

```swift
return Task(
    title: String(localized: "TASK_SOCIAL_SUPPORT_QUESTIONNAIRE_TITLE"),
    description: String(localized: "TASK_SOCIAL_SUPPORT_QUESTIONNAIRE_DESCRIPTION"),
    schedule: Schedule(
        start: Calendar.current.startOfDay(for: Date()),
        repetition: .matching(dateComponents),
        end: .numberOfEvents(365)
    ),
    notifications: true,
    context: BehaviorTaskContext.questionnaire(Bundle.main.questionnaire(withName: "SocialSupportQuestionnaire"))
)
```

As you can see, you can customize the title, description, and schedule for the particular task.

Every [Task](https://swiftpackageindex.com/stanfordspezi/spezischeduler/documentation/spezischeduler/task) in SpeziScheduler also has a customized `context` attached to it. 

On line 41, we are attaching a context which contains our questionnaire, drawn from a JSON file in the Resources folder called "SocialSupportQuestionnaire.json". 

```swift
 context: BehaviorTaskContext.questionnaire(Bundle.main.questionnaire(withName: "SocialSupportQuestionnaire"))
```

In [ScheduleView](https://github.com/CS342/2024-Behavior/blob/main/Behavior/Schedule/ScheduleView.swift), when the user taps to start a task, we look at the context and decide what should happen depending on what's attached. If it's a questionnaire, we then launch it using the [SpeziQuestionnaire](https://github.com/StanfordSpezi/SpeziQuestionnaire) module.

If you place your questionnaire file (e.g. "Questionnaire.json") in the `Behavior/Resources` folder, then replace the string "SocialSupportQuestionnaire" in the line above with the name of your questionnaire, it will be rendered instead. (You do not need to include the ".json" file ending because [it is assumed](https://github.com/CS342/2024-Behavior/blob/main/Behavior/Schedule/Bundle%2BQuestionnaire.swift).)

**Reply by carolinentran** [reply]
Thank you so much for your help!

**Comment by terrylin8** [comment]
Hi! I am having some trouble with the task as well. I have replaced the json file, modified the task in TBIScheduler accordingly, and the updated task showing my new questionnaire name is displayed in EventContextView. However, this update is not reflected in ScheduleView, as the scheduled task is still the original questionnaire. I've spent hours and can't seem to update ScheduleView. I am now thinking maybe calculateEventContextsByDate needs to be changed in order to propagate the change in TBIScheduler and EvnentContext over. Any suggestions on how to approach it?

**Reply by carolinentran** [reply]
I've also been working on this for a while and it's not updating for me neither. I'll let you know if I figure something out.

**Reply by PSchmiedmayer** [reply]
@carolinentran @twtl67 It might be that you need to re-install the application; you can un-install it on the iOS Simulator by long-pressing the app icon on the simulator Home Screen and deleting the app as you would do on an iPhone: https://support.apple.com/guide/iphone/remove-apps-iph248b543ca/ios

The Spezi Scheduler module is built in a way that it instantiates all the tasks defined in the `Configuration` that a user has to perform the first time the app launches and then loads them from disk every time after that. The reason is that you don't want to reset and override all tasks every time you launch the app as a user. This would reset all progress and override everything that you have done so far.

Therefore adding a new task after a user has already installed the app (e.g. in an update) [needs to be done using the following API](https://swiftpackageindex.com/stanfordspezi/spezischeduler/0.8.0/documentation/spezischeduler/scheduler/schedule(task:) or you need to completely reset the Scheduler module (by removing all tasks).
As your addition is like installing an update, the easiest way during development is just to uninstall and re-install the app.

Sometimes (e.g., when you develop an app that stores data in secure storage), completely resetting the Simulator also helps a lot as it clears all secret information stored in an app. You can do this in the menu bar in the iOS Simulator using _"Device" > "Erase All Contents and Settings..."_.

**Reply by carolinentran** [reply]
Hi Paul, 

Thank you so much! It actually works and updates when I run the simulator (without having to delete and re-install). I think the preview on xcode was having issues so I'll try deleting and re-installing to see if that fixes it on the xcode preview.

**Reply by PSchmiedmayer** [reply]
Great to hear! Xcode previews might be a different thing, you might need to just restart Xcode for that.
